### ISSUE TITLE
observe() callback is called for a change in another Store

### ISSUE THREAD

**Issue by mindjuice** [optimizely/nuclear-js#27]
I have run into a problem where one of my Nuclear app ``observe`` callbacks is being called when a different Store is changed than the one it is observing.

My ``server-api.js`` file has the following ``observe`` call:

    initialize: function() {
      reactor.observe('query.query', sendQuery);
    }

The ``sendQuery`` function is defined like this:

    let sendQuery = (query) => {
      console.log("Issuing server request!");
      Actions.setRequestInProgress(true);
      // etc.
      // Then in an async callback
        Actions.setRequestInProgress(false);
    };

My DataStore has the following code to register a listener for ``setRequestInProgress`` (the QueryStore does not register a listener for ``setRequestInProgress``):

    this.on(Const.setRequestInProgress, (state, isInProgress) => {
      console.log('data-store: set isInProgress to ' + isInProgress);
      return state.set('requestInProgress', isInProgress);
    });

I have a button that triggers an action that changes ``query.query``.  As expected, this triggers my ``sendQuery`` method since it was observing ``query.query``.

The ``Actions.setRequestInProgress(true)`` action triggers a change in a the DataStore (in the ``data. requestInProgress`` keypath) so I can keep track of whether there is a query in progress.

The problem is that while handling the data change for ``setRequestInProgress``, my ``sendQuery`` function gets called again, even though nothing new changed in ``query.query``.  It's as if it gets called for any data change if there is already a change in progress from an earlier Action.

I was able to work around this by changing my actions in ``sendQuery`` to use ``setTimeout`` like so:
    
    setTimeout(Actions.setRequestInProgress.bind(null, true), 0);

It seems to me that this should not be necessary though.

Am I doing something weird/wrong or is this a defect?

Thanks!

Ken

**Comment by jordangarcia** [comment]
Hi Ken!

This is a pattern that Flux tries to avoid and NuclearJS explicitly will throw an error if you dispatch within a dispatch or observe.  

The main reasoning here is to keep app with a large amount of state understandable is to not cause cascading dispatches.  A single action dispatched should cause the AppState to go from S1 to S2.  If we allow cascading actions to be dispatched than a single action can cause AppState to go from S1 to S2 to S3 to S4 and so on, making it very hard to conceptually understand what your system is doing and make bug tracking much harder.

The proper way to do this in Nuclear is to do it all in an action.

For instance:

```
function sendQuery(query) {
  flux.dispatch('SET_QUERY', { query: query });
  setRequestInProgress(true);
  // make ajax request
  setRequestInProgress(false);
}
```

The way prevents side effects from happening in response to state changing.  The frequency at which app state can change can be very high and also unpredictable, so having expensive side-effect such as a ASYNC request can cause poor performance.

"Actions" as they are used in nuclear do not have to be 1-1 with a dispatch they can contain multiple dispatches.  They represent a semantic operation happening on your system and should be implemented in that way.

Hope this helps!

**Comment by mindjuice** [comment]
Hi Jordan,

Thanks for the explanation.  I thought it might be something like that, since I've run into a similar problem with vanilla flux on my last project.  I have a few follow-ups to make sure I'm following if that's OK (apologies if I am missing something obvious here).

1. In your code example, you call ``flux.dispatch()``, and then ``setRequestInProgress(true)``.  My assumption is that ``setRequestInProgress()`` is also an action, is that right?  If so, then are you saying that doing dispatches serially like that is okay?  I also assume that ``dispatch()`` doesn't do anything async (although a listening Store might), correct?

2. It seems pretty natural to want to listen for when some state changes and perform some action in response ('if this then do that').  If all state changes should be driven by dispatching an action, but you can't call ``dispatch()`` from an ``observe()`` callback then what **can** you do from an ``observe()`` callback?

 Putting the ``flux.dispatch()`` call into the same action with some code that happens to want to do something in response to that state change (i.e., set a ``requestInProgress`` flag and send an async request) seems like unnecessary coupling to me.

 What if you also wanted to save each query to a log file?  Would you then make the method ``sendQueryAndLog()`` and add more serial dispatches to the end?  This would quickly become a Frankenstein function.

 I don't want my UI to directly trigger a ``sendQuery()`` action.  I want it to trigger actions that modify some state that happens to factor into making a query.  For example, I might want to trigger an action like ``setQueryDateRange()``.  The code on the other end of the ``observe()`` can decide if the updated state requires a new request to be sent or not.  Maybe the new date range is a subset of dates already received, so no query is required, or perhaps there is a debounce interval that hasn't expired yet so we avoid making requests too quickly.

  From what I can tell (please correct me if I'm wrong), in my code, we get a sequence of actions like this:
  * dispatch ``SET_QUERY`` action
  * Nuclear calls QueryStore listener since it registered for that action
  * QueryStore listener updates its immutable state
  * Nuclear calls my ``observe()`` callback
  * ``observe()`` callback dispatches ``SET_REQUEST_IN_PROGRESS`` action
  * Nuclear calls DataStore listener since it registered for that action
  * DataStore listener updates its immutable state
  * Nobody is observing DataStore, so the dispatch is done
  * ``observe()`` callback initiates an async request to be resolved later
  * We unwind our call stack and then the original dispatch ``SET_QUERY`` is now done

 With your ``sendQuery()``, I expect we get something more like: 
  * dispatch ``SET_QUERY`` action
  * Nuclear calls QueryStore listener since it registered for that action
  * QueryStore listener updates its immutable state
  * Nobody is observing QueryStore, so the dispatch ``SET_QUERY`` is now done
  * ``setRequestInProgress(true)`` gets called from ``sendQuery()``
  * It dispatches ``SET_REQUEST_IN_PROGRESS`` action
  * Nuclear calls DataStore listener since it registered for that action
  * DataStore listener updates its immutable state
  * Nobody is observing DataStore, so the dispatch is done
  * ``sendQuery()`` then initiates an async request to be resolved later

 In the end, it seems that both both versions cause the same state to be modified and initiate an async request, but in a different order.  If I understand you correctly, you believe that this alternate order is better, but I'm not following why.  As long as we don't make loops, it seems to be the same result to me, but without the unnecessary coupling in ``sendQuery()``.

3. You mentioned that Nuclear will explicitly throw an error if I dispatch at the wrong time, but I wasn't seeing that.  I just saw my ``observe()`` callback get called when it shouldn't have been.

Thanks Jordan.

**Comment by jordangarcia** [comment]
Hi Ken,

You make a lot of very good points regarding the equivalency of both techniques.  It seems like your main concern is regarding the coupling of your actions together, which is a valid concern.

> For example, I might want to trigger an action like setQueryDateRange(). The code on the other end of the observe() can decide if the updated state requires a new request to be sent or not.

I agree with this mentality of wanting to do something and then having another piece of the system be the one deciding whether to perform some action op top of it.  However the issue with doing this in an observe is you often want to know the previous state and the current state to of something to make this decision.  Being in an observe handler only grants you access to the current state of the system after it has changed.

Doing this decision making process in an action allows you to write code that performs the following checks:

1.  Check the current state of the system
2.  Dispatch an action that causes state change synchronously (this part is key)
3.  Check the new state of the system while having the previous state in context and then making a decision to do something else.

As to not coupling everything into one action, it's quite common for actions to compose other actions.  In our codebase at Optimizely we have several modules that we consider "low level" or model layer modules and then we have implementation layer modules that can couple things together because they are tied to a specific implementation.  The beauty of modules having actions / getters is a higher level module can compose lower level modules without coupling your lower level modules together and thus keeping them reusable.

Hope this helps.  I understand the propensity to want to get in the pattern of reactive programming for your entire system.  I do thing at a point it makes the system harder to understand and hard to refactor.

**Comment by mindjuice** [comment]
Thanks for the additional detail.  I'll give it a try your way on Monday and see how it goes. :-)

**Comment by bhamodi** [comment]
@mindjuice - can we close out this issue?

**Comment by mindjuice** [comment]
Yes, I think that the fix discussed in #130 and #131 would have helped with this situation too, as I was calling ``dispatch()`` from an ``observe()`` callback.
